### PR TITLE
Use schema definition for `drop_table` operation

### PR DESCRIPTION
**Pull request by adrianna-chang-shopify**
### Summary

This PR is part of a larger effort to refactor the Rails migration workflow. The goal is to decouple the production of DDL from its execution, which will eventually allow applications to customize the objects responsible for producing and executing DDL.

The first step in this refactor is to ensure that all schema alterations can be represented by a schema definition, and that `SchemaCreation` objects can visit these definitions to produce SQL / DDL. Some operations (like `#create_table`) are already doing this. The intent here is to refactor all operations to follow this pattern.

This PR introduces a schema definition, `DropTableDefinition` , for dropping a table from a database. It defines `visit_DropTableDefinition` methods on the relevant `SchemaCreation` objects, and extracts the SQL creation from the adapters to these methods.

I'm relying on existing test coverage to validate that this refactor works as expected.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb [hunk 1]
```diff
@@ -9,6 +9,13 @@ def visit_AlterTable(o)
             super << o.constraint_validations.map { |fk| visit_ValidateConstraint fk }.join(" ")
           end
 
+          def visit_DropTableDefinition(o)
+            sql = "DROP TABLE "
+            sql += "IF EXISTS " if o.if_exists
+            sql + quote_table_name(o.name)
```
Comment (gmcgibbon) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:15]:
```suggestion
            sql += quote_table_name(o.name)
```
or `<<` should be added here right?

**Comment by adrianna-chang-shopify** [comment]
cc @rafaelfranca @paarthmadan @edward

**Comment by matthewd** [comment]
This feels like it's going to pretty substantially bulk up these files, and the indirection involved, by the time it's chased through the rest of the DDL. 

Could we explore moving the methods as-is into a strategy class (with straight delegation from the connection) instead? 
I suspect that might allow a similar degree of downstream flexibility without needing the extra classes per operation (and in a bunch of cases, per adapter per operation).

To me, the object on `create_table` and friends is an unfortunate reflection of their much higher complexity.. despite the structural simplicity of mirroring it forward, IMO we'd be better off if we can achieve that same goals without doing so.

**Comment by adrianna-chang-shopify** [comment]
Hi @matthewd, thanks for the feedback! I understand your concerns about the additional bulk / indirection related to making all operations conform to the existing visitor pattern that some schema statements are using.

> Could we explore moving the methods as-is into a strategy class

We explored a solution similar to this, albeit with a custom adapter rather than using strategy pattern to swap out the behaviour. A bit more about our use case at Shopify:
- Shopify uses a service outside of Rails to run migrations across our applications. It consumes operations as JSON, and handles performing these schema alterations in horizontally-sharded environments using an LHM backend.
- To achieve this today, we are monkey-patching the `#execute` method on the `MySql2Adapter`, parsing the SQL, turning that into JSON, and submitting that to the schema migrations service.

Our long term vision with schema definitions is this:
- If all schema operations have a corresponding schema definition upstream in Rails, then:
- We could write our own translation object, which would consume a schema definition and produce JSON (just as `SchemaCreation` objects in Rails today consume definitions and produce SQL strings)
- If we could decouple SQL production and execution by moving all SQL production to `SchemaCreation` objects upstream, then:
- We could write our own execution object, which would pass this JSON along to the schema migrations service

All of this without monkey patching Rails. The really fundamental piece on our end is this: **schema definitions would give us sufficient information to build the JSON payload our schema migrations service needs**. I think a strategy approach is insufficient for our needs, because we don't just want to swap out the behaviour for something like `#create_table` -- we want Rails to keep doing the work of turning `#create_table` into a `TableDefinition`, and then we want to translate this definition into our own DDL (JSON) instead of turning it into SQL and executing it.

Let me know if you think your ideas around using a strategy class could work with what I've described! 😄

**Pull request closed without merging**